File: Server/data/models/Page.ts
import {PageElement} from './PageElement';

class Page {
    id: number;
    title: string;
    section: string;
    elements: Array<PageElement>;

    private elementIdCounter: number;
    constructor(id: number = 0, title: string = 'Untitled Page', elements: Array<PageElement> = [])
    {
        this.id = id;
        this.title = title;
        this.elements = elements;
        let maxNumber = 0;
        elements.forEach(element => {
            if (element.id > maxNumber)
                maxNumber = element.id;
        });
        this.elementIdCounter = maxNumber + 1;
    }

    public appendNewElement()
    {
        let id = this.getNextElementId();
        let element = new PageElement(id);
        this.elements.push(element);
    }

    public addElement(type: string = 'p', content: string = '')
    {
        const id = this.getNextElementId();
        let element = new PageElement(id, type, content);
        this.elements.push(element);
    }
    public getElement(elementId: number): PageElement
    {
        return this.elements.find( element => element.id == elementId);
    }

    public updateElement(elementId: number, content: string)
    {
        let element = this.elements.find( element => element.id == elementId);
        element.content = content;
    }

    public deleteElement(elementId: number)
    {
        throw new Error('deleteElement not implemented yet');
    }

    private getNextElementId(): number
    {
        return ++this.elementIdCounter;
    }


}

export {Page};

File: Notes/Data/Models/Page.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Notes.Data.Models
{
    public class Page
    {
        public int Id { get; set; }
        public string Title { get; set; }
        public List<PageElement> Elements { get; private set; }

        private IdCounter elementIdCounter;
        public Page(int id = 0, string title = "Untitled Page", List<PageElement> elements = null)
        {
            Id = id;
            Title = title;
            Elements = elements ?? new List<PageElement>();

            int maxElementId = 0;
            Elements.ForEach(element =>
            {
                if (element.Id > maxElementId)
                    maxElementId = element.Id;
            });
            elementIdCounter = new IdCounter(maxElementId + 1);
        }

        public void appendElement()
        {
            PageElement element = new PageElement();
            element.Id = elementIdCounter.getNextThenIncrement();
        }
        public void addElement(PageElement element)
        {
            int id = elementIdCounter.getNextThenIncrement();
            element.Id = id;
            Elements.Add(element);
        }

        public void addElements(params PageElement[] elements)
        {
            foreach (PageElement element in elements)
            {
                this.addElement(element);
            }
        }

        public PageElement getElement(int elementId)
        {
            return Elements.Find(element => element.Id == elementId);
        }

        public void updateElement(int elementId, string content)
        {
            throw new NotImplementedException();
        }

        public void deleteElement(int elementId)
        {
            throw new NotImplementedException();
        }

    }
}